### PR TITLE
Add List/Scan/Count Workflow APIs

### PR DESCRIPTION
**Pull request by vancexu**
This PR add 3 visibility APIs for SQL like query with ElasticSearch.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -220,6 +220,37 @@ type (
 		//  - EntityNotExistError
 		ListOpenWorkflow(ctx context.Context, request *s.ListOpenWorkflowExecutionsRequest) (*s.ListOpenWorkflowExecutionsResponse, error)
 
+		// ListWorkflow gets workflow executions based on query. This API only works with ElasticSearch,
+		// and will return BadRequestError when using Cassandra or MySQL. The query is basically the SQL WHERE clause,
+		// examples:
+		//  - "(WorkflowID = 'wid1' or (WorkflowType = 'type2' and WorkflowID = 'wid2'))".
+		//  - "CloseTime between '2019-08-27T15:04:05+00:00' and '2019-08-28T15:04:05+00:00'".
+		//  - to list only open workflow use "CloseTime = missing"
+		// Retrieved workflow executions are sorted by StartTime in descending order when list open workflow,
```
Comment (shreyassrivatsan) [inline client/client.go]: `when listing open workflows, and sorted by CloseTime in descending order for other queries`

File: client/client.go [hunk 2]
```diff
@@ -220,6 +220,37 @@ type (
...
+
+		// ScanWorkflow gets workflow executions based on query. This API only works with ElasticSearch,
+		// and will return BadRequestError when using Cassandra or MySQL. The query is basically the SQL WHERE clause
+		// (see ListWorkflow for query examples).
+		// ScanWorkflow is similar to ListWorkflow, but it is more efficient to retrieve large amount of workflows,
+		// and the result is not sorted.
+		// The errors it can return:
+		//  - BadRequestError
+		//  - InternalServiceError
+		ScanWorkflow(ctx context.Context, request *s.ListWorkflowExecutionsRequest) (*s.ListWorkflowExecutionsResponse, error)
```
Comment (shreyassrivatsan) [inline client/client.go]: How do users know when to use what? Can we just make sorting an option on the request so and say that sorting will be expensive.. Seems like we should do that in the backend and only expose one API actually..

File: client/client.go [hunk 3]
```diff
@@ -220,6 +220,37 @@ type (
...
+
+		// CountWorkflow gets number of workflow executions based on query. This API only works with ElasticSearch,
+		// and will return BadRequestError when using Cassandra or MySQL. The query is basically the SQL WHERE clause
```
Comment (shreyassrivatsan) [inline client/client.go]: All 3 descriptions say the query is basically the `SQL WHERE clause`, but one is with sorting, one just a count and one a simple where.. I understand `COUNT` is more performant than doing a list and counting on the client side, but was there a specific request for count as an API?

File: client/client.go (hunk 1 above)
Comment (vancexu) [reply client/client.go]: Updated.

File: client/client.go (hunk 2 above)
Comment (vancexu) [reply client/client.go]: Good question, I update comment to be more specific. 

It is not recommended by ES to use scroll for all search requests without order, so I have separate scan API.

> Scrolling is not intended for real time user requests, but rather for processing large amounts of data, e.g. in order to reindex the contents of one index into a new index with a different configuration.
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html

File: client/client.go (hunk 3 above)
Comment (vancexu) [reply client/client.go]: Yes, so far I know 2 use case:
- Cadence Report 
- CLI. Oncall asked me multiple times how many open workflows luna has during some time range.

File: internal/client.go [hunk 4]
```diff
@@ -205,6 +205,37 @@ type (
...
 		ListOpenWorkflow(ctx context.Context, request *s.ListOpenWorkflowExecutionsRequest) (*s.ListOpenWorkflowExecutionsResponse, error)
 
+		// ListWorkflow gets workflow executions based on query. This API only works with ElasticSearch,
+		// and will return BadRequestError when using Cassandra or MySQL. The query is basically the SQL WHERE clause,
+		// examples:
+		//  - "(WorkflowID = 'wid1' or (WorkflowType = 'type2' and WorkflowID = 'wid2'))".
+		//  - "CloseTime between '2019-08-27T15:04:05+00:00' and '2019-08-28T15:04:05+00:00'".
+		//  - to list only open workflow use "CloseTime = missing"
+		// Retrieved workflow executions are sorted by StartTime in descending order when list open workflow,
+		// are sorted by CloseTime in descending order for other query.
```
Comment (shreyassrivatsan) [inline internal/client.go]: same change as in other file.

File: internal/client.go [hunk 5]
```diff
@@ -205,6 +205,37 @@ type (
...
+		// The errors it can return:
+		//  - BadRequestError
+		//  - InternalServiceError
+		ListWorkflow(ctx context.Context, request *s.ListWorkflowExecutionsRequest) (*s.ListWorkflowExecutionsResponse, error)
+
+		// ScanWorkflow gets workflow executions based on query. This API only works with ElasticSearch,
+		// and will return BadRequestError when using Cassandra or MySQL. The query is basically the SQL WHERE clause
+		// (see ListWorkflow for query examples).
+		// ScanWorkflow is similar to ListWorkflow, but it is more efficient to retrieve large amount of workflows,
+		// and the result is not sorted.
```
Comment (shreyassrivatsan) [inline internal/client.go]: same as above file

**Comment by shreyassrivatsan** [comment]
Looks good, just a couple of `comment` comments..

**Pull request merged**